src/solver/mod.rs

```diff
@@ -167,6 +172,78 @@ impl<D: DependencyProvider, RT: AsyncRuntime> Solver<D, RT> {
         root_requirements: Vec<VersionSetId>,
         root_constraints: Vec<VersionSetId>,
     ) -> Result<Vec<SolvableId>, UnsolvableOrCancelled> {
+        self.solve_for_root(root_requirements, root_constraints)?;
+
+        let steps: Vec<_> = self.chosen_solvables().collect();
+
+        tracing::trace!("Solvables found:");
+        for step in &steps {
+            tracing::trace!(
+                " - {}",
+                InternalSolvableId::from(*step).display(self.provider())
+            );
+        }
+
+        Ok(steps)
+    }
+
+    /// Solves for the provided `root_requirements` and `root_constraints`, along with
+    /// a set of additional (optional) solvables.
+    ///
+    /// The `root_requirements` are packages that must be included in the
+    /// solution. `root_constraints` are additional constraints applicable to
+    /// individual packages that the solvable (if any) chosen for that package
+    /// must adhere to.
+    ///
+    /// The solver first solves for the root requirements and constraints, and then
+    /// tries to include in the solution as many of the additional solvables as it can.
+    /// Each additional solvable is subject to all the clauses and decisions introduced
+    /// for all the previously decided solvables in the solution.
+    ///
+    /// Unless the corresponding package has been requested by a version set in another
+    /// solvable's clauses, each additional solvable is _not_ subject to the
+    /// package-level clauses introduced in [`DependencyProvider::get_candidates`] since the
+    /// solvables have been requested specifically (not through a version set) in the solution.
+    ///
+    /// Returns [`UnsolvableOrCancelled::Unsolvable`] only if the _root_ requirements and
+    /// constraints could not be solved for; if an additional solvable is unsolvable, it is
+    /// simply not included in the solution.
+    pub fn solve_with_additional(
+        &mut self,
+        root_requirements: Vec<VersionSetId>,
+        root_constraints: Vec<VersionSetId>,
+        additional_solvables: impl IntoIterator<Item = SolvableId>,
+    ) -> Result<impl Iterator<Item = SolvableId> + '_, UnsolvableOrCancelled> {
```

Comment:
Instead of having a separate function, I propose we break the API and introduce a struct over multiple arguments. Otherwise, I'm a bit worried that in the future we will have more options and have to add more "overloads". 

Ideally, this struct implements the `Default` trait. So you can write something like:

```rust
solver.solve(SolveGoal { root_requirements: vec![..], ..SolveGoal::default() })
```

We can also rename `root_requirements` to `requirements`, the same for `constraints`. I would also like to rename `additional_solvables` to `soft_requirements`. In the future, we will have to break the API again when we change from solvables to requirements but I think the name `requirements` conveys better what the intent is. In the doc comment for that field we should be very clear that we know currently it only accepts solvables and we will most likely change that in the future.

Would you be able to make these changes yourself?

Reply:
I should be able to, yes. Would it make sense to have `SolveGoal` (or the equivalent) to have private fields and provide constructor methods/builders to make future additions non-breaking changes, or should we just keep all fields public and allow API breakages when adding a new argument (with the assumption that people using `..SolveGoal::default()` will be unaffected?

Reply:
Let's go with a struct for now because that makes the C++ implementation simpler.

Reply:
I think `Problem` would be a good name for this input struct (since it defines the problem the solver is trying to solve for). Would it make sense to rename the existing `Problem` structs in the `solver::problem` submodule to something a bit more relevant, such as `Conflict`, since they describe the nature of an unsatisfiable conflict, rather than the problem being solved for itself?

Reply:
Sounds good!